bin/genome-env-next

```diff
@@ -0,0 +1,432 @@
+    fi
+
+    if test -z "$WF_REPO"
+    then
+        if test -z "$TOP" -o ! -d "$TOP/workflow"
+        then
+            err "-w must be specified if not in a Genome repo"
+            exit 2
+        fi
+        SUBMODULES=("${SUBMODULES[@]}" workflow)
+        WF_REPO="$TOP/workflow"
+    fi
+
+    if test -n "$SUBMODULES"
+    then
+        prepare_submodules "${SUBMODULES[@]}"
+    fi
+}
+
+function __quiet_git {
+    (
+        set +o errexit
+        OUT="$(git "$@" 2>&1)"
+        GIT_EXIT=$?
+        if test $GIT_EXIT -ne 0
+        then
+            echo -e "$OUT" 1>&2
+            exit $GIT_EXIT
+        fi
+    )
+}
+
+function stderr {
+    echo "$@" 1>&2
+}
+
+function prepare_submodules {
+    (
+        stderr -e "\n=> Preparing Submodules..."
+        cd "$(git rev-parse --show-cdup)"
+
+        __quiet_git submodule sync "$@"
+
+        for submodule in "$@"; do
+            stderr -n "   $submodule: "
+            if git submodule status $submodule | grep -q ^-
+            then
+                __quiet_git submodule update --init $submodule
+                stderr "OK"
+            else
+                (
+                    EXPECTED_HEAD=$(git ls-tree HEAD $submodule | awk '{print $3}')
+                    cd $submodule
+                    HEAD=$(git rev-parse HEAD)
+
+                    if ! git cat-file -e $EXPECTED_HEAD
+                    then
+                        __quiet_git fetch --all
+                    fi
+
+                    if ! git status --porcelain | grep -q . \
+                        && git merge-base --is-ancestor $HEAD $EXPECTED_HEAD
+                    then
+                        cd - 1> /dev/null
+                        __quiet_git submodule update --init $submodule
+                        stderr "OK"
+                    else
+                        stderr "ERROR: Not updating '$submodule' submodule because it has local changes!"
+                        exit 1;
+                    fi
+                )
+            fi
+        done
+    )
+}
+
+function regen_db {
+    local DB_TXT="$1"
+    echo -n '.' 1>&2
+    DB_BIN=${DB_TXT/-dump/}
+    if [[ "$DB_BIN" -ot "$DB_TXT" ]]; then
+        rm -f "$DB_BIN"
+        sqlite3 "$DB_BIN" < "$DB_TXT"
+    fi
+}
+
+function rebuild_meta_db {
+    echo -en "\n=> Rebuilding Meta DB..." 1>&2
+
+    if git rev-parse --is-inside-work-tree 1> /dev/null 2> /dev/null
+    then
+        # ls-files is fast but then we have to also query submodules
+        # foreach defines $name which serves to prefix the submodule path
+        ( git ls-files \
+            && git submodule foreach "git ls-files | sed \"s|^|\$name/|\"" ) \
+        | grep -P '\.sqlite3n?-dump$' \
+        | while read DB_TXT
+        do
+            regen_db "$DB_TXT"
+        done
+    else
+        find "$PWD" -iname '*sqlite3-dump' -or -iname '*sqlite3n-dump' | while read DB_TXT; do
+            regen_db "$DB_TXT"
+        done
+    fi
+
+    echo ' done' 1>&2
```

Comment:
I'll comment on this only here, but I think a lot of the locations that match a search for >&2 could use the stderr function you defined instead.